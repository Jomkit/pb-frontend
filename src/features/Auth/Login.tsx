import Frame from '../../components/Frame'

const Login = ({children}: {children: any}) => {
  return (
    <Frame bgColor='bg-green-400' title='Login'>
      {children}
    </Frame>
  )
}

export default Login