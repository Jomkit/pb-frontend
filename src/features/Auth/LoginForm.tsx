
import * as Yup from 'yup';
import DynamicForm from '../../components/DynamicForm';

const loginSchema = Yup.object().shape({
  username: Yup.string().required(),
  password: Yup.string().required()
})

const LoginForm = ({login}: {login: (username: string, password: string) => void}) => {

  const handleSubmit = async (values: any, resetForm: Function) => {
    try{
      await login(values.username, values.password);

    }catch(err: any){
      console.log(err[0] || err);
      
      alert(err[0] || err);
      resetForm();
    }
  }

  return (
    <DynamicForm initialValues={{ username: '', password: '' }} schema={loginSchema} handleSubmit={handleSubmit} />
  )
}

export default LoginForm