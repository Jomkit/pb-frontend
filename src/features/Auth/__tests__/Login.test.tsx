import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Login from "../Login";
import LoginForm from "../LoginForm";
import { ReactNode } from "react";

describe("Login", function() {
    const mockedLogin = vi.fn();

    function setup(tsx: ReactNode) {
        return {
            user: userEvent.setup(),
            ...render(tsx)
        }
    }
    
    it('should render without crashing', () => { 
        render(<Login>
            <LoginForm login={mockedLogin} />
        </Login>);
    })
    it('should match snapshot', () => { 
        const { asFragment } = render(
            <Login>
                <LoginForm login={mockedLogin} />
            </Login>
        )

        expect(asFragment()).toMatchSnapshot();
    })

    it('should call login', async () => { 
        const { user, getByRole, getByLabelText } = setup(
            <Login>
                <LoginForm login={mockedLogin} />
            </Login>
        )

        const usernameField = getByLabelText("username");
        const passwordField = getByLabelText("password");
        await user.type(usernameField, "testUsername");
        await user.type(passwordField, "testPassword");

        const submitBtn = getByRole('button', {name: /Submit/i})
        user.click(submitBtn);
        await waitFor(() => {
            expect(mockedLogin).toHaveBeenCalledWith("testUsername", "testPassword");
            expect(mockedLogin).toHaveBeenCalled();
        })
    })

    it("should not call login if username or password is missing", async () => {
        const { user, getByRole, findByLabelText } = setup(
            <Login>
                <LoginForm login={mockedLogin} />
            </Login>
        )

        const usernameField = await findByLabelText("username");
        const passwordField = await findByLabelText("password");
        await user.type(usernameField, "testUsername");
        await user.type(passwordField, "[Tab]");

        const submitBtn = getByRole('button', {name: /Submit/i})
        
        try{
            await user.click(submitBtn);
        }catch(err: any){
            expect(err).toBeTruthy();
            expect(err.message).toContain("Password is a required field");
        }
    })
})