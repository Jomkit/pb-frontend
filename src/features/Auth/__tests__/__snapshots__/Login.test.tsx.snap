// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > should match snapshot 1`] = `
<DocumentFragment>
  <section
    class="text-center items-start min-h-screen w-screen p-4 bg-green-400"
  >
    <h1>
      Login
    </h1>
    <div
      class="p-10 justify-center"
    >
      <form
        action="#"
      >
        <div
          class="rounded-lg border border-gray-200 bg-white shadow-md dark:border-gray-700 dark:bg-gray-800 flex flex-col text-start w-1/2 m-auto"
          data-testid="flowbite-card"
        >
          <div
            class="flex h-full flex-col justify-center gap-4 p-6"
          >
            <label
              class="block text-gray-700 text-sm font-bold mb-2"
              for="username"
            >
              username
            </label>
            <input
              class="form-field1 focus:outline-none focus:shadow-outline"
              id="username"
              name="username"
              type="text"
              value=""
            />
            <label
              class="block text-gray-700 text-sm font-bold mb-2"
              for="password"
            >
              password
            </label>
            <input
              class="form-field1 focus:outline-none focus:shadow-outline"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </div>
        </div>
        <button
          class="btn bg-blue-700 text-white"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </section>
</DocumentFragment>
`;
